### ISSUE TITLE
FEATURE: need to check math dimensions

### ISSUE THREAD

**Issue by buggsley** [MyersResearchGroup/iBioSim#176]
New a method to calculate the dimensions of math.  It should throw exceptions when:

1) A selector has too many selected dimensions for the object.
2) When there is a ragged vector.
3) When the dimension of two objects in a multiple operator function do not match.

This function should return the dimensions of the math allowing checks against the LHS of assignments.

Check run for:

1) Initial assignments
2) Rules
3) Event assignments
4) Kinetic laws / compared with dimensions of the reaction
5) Triggers/delays/priorities with dimensions of the event
6) Constraints with their dimension
7) Index math must be scalar.

**Comment by buggsley** [comment]
I have a method called checkVectorMath(model, mathContainer). MathContainers are
math wrappers such as 
assignment rules, kinetic law, event assignment, etc. This is checking is the
math operation is valid, where vector 
should match in size.

**Comment by buggsley** [comment]
I'm considering the bug report about checking math dimensions.  For this, you
create a method called 
checkVectorMath to return the dimensions of a math formula, so it could be
compared against the dimensions of 
the variable or other structure that it is associated with.  However, I'm having
trouble coming up with a situation 
where this is needed.  I think when an assignment is made to all the index
variables that all math formulas must 
reduce to scalar formulas unless I'm missing something.  For example, assume the
math is a kinetic law.  In this 
case, you select a value for each dimension id for the reaction, and when you
plug this into the math for the 
kinetic law, it must return a single scalar value.  So, I think as long as the
references in the math do not go out-
of-bounds for some allowed value, then this formula is going to be okay.  In all
cases that we listed, I think they 
same argument applies whether a rule, event assignment, event trigger, etc.  Am
I missing something?  

Chris

**Comment by buggsley** [comment]
Hi Chris,

The method checkVectorMath returns a boolean and tells if a vector is ragged or
not. 

We agreed that event assignment, rules, initial assignment, etc must have index
objects for array operations so 
they all return a scalar value. I think the only case where we need to check if
the vector size matches the 
dimensions of an object is for event. Trigger, Priority, and Delay are not
allowed to have ListtOfDimensions, but 
they can be vectors as far as I remember. 

I don’t remember where these checks are called. I remember only checking if
the math is statically computable 
and array bounds.

Leandro

**Comment by buggsley** [comment]
The method checkVectorMath returns a boolean and tells if a vector is ragged or
not. 

We agreed that event assignment, rules, initial assignment, etc must have index
objects for array operations so 
they all return a scalar value. I think the only case where we need to check if
the vector size matches the 
dimensions of an object is for event. Trigger, Priority, and Delay are not
allowed to have ListtOfDimensions, but 
they can be vectors as far as I remember. 

I think even trigger math that is a vector, we require it to have an index into
the vector likely using the event's 
dimension id in order to turn it into a scalar.   

I don’t remember where these checks are called. I remember only checking if
the math is statically computable 
and array bounds.

Not sure if those checks would catch a ragged vector or not.  Does seem to me
though that all math has to reduce 
to a scalar now, since we are not allowing arrays to be used directly.  This may
imply some simplifications to the 
validation rules, I think.

Chris

**Comment by buggsley** [comment]
Scott: you can now add use this function to check math dimensions against the dimensions of the object they are associated with.  We can discuss this one on Wednesday.

**Comment by cjmyers** [comment]
Leandro can you have a look at this one to see if it is still a problem.

**Comment by leandrohw** [comment]
1. A selector has too many selected dimensions for the object.
  - This seems to work. However, this does not work for vectors. For instance, I can assign {{1}}[0] to a one dimensional array.

2. When there is a ragged vector.
- This shouldn't be a problem.

3. When the dimension of two objects in a multiple operator function do not match.
- No longer applicable. All elements should reduce to scalars so no vector operation is valid.

**Comment by cjmyers** [comment]
Yes, there does seems to be some issues with vector initial assignments.  Going to move this to 3.1.
